README.md
# Routex
Simple and fast URL router powered by regex with predefined patterns.

### Introduction
This router has been developed due to the need of a very simple and lightweight router. Unlike routers used in (micro-)frameworks, this class is just one file. It offers powerful routing using PCRE compliant regex, or simply use the built in placeholders.

### Requirements
- PHP 7.4+
- URL rewriting (see .htaccess)

### Usage
```php
<?php

require_once __DIR__.'/Hybula/Routex/Routex.php';
use Hybula\Routex\Routex;

// Override the default 404, you may set a different status code too. If not set run() will return false and a 404 status code;
Routex::error(404, function(){
    echo '404';
});

// Enable basic debugging by printing pattern and routes;
Routex::debug(true);

// Change the base URL, use this if you are routing inside a directory instead of root;
Routex::base('/v1/');

// Load build in patterns and set optional custom patterns to be used in routing URLs later;
Routex::patterns([
    ':hostname' => '((?!\-)(?:(?:[a-zA-Z\d][a-zA-Z\d\-]{0,61})?[a-zA-Z\d]\.){1,126}(?!\d+)[a-zA-Z\d]{1,63})'
]);

// Load external files for routing, so you can create seperate files per route;
Routex::load(__DIR__.'/routes');

// Example how to route the index;
Routex::get('/', function(){
    echo 'Hello world!';
});

// Example on how to use PCRE regex to route and capture params;
Routex::get('blog/(\w+)/(\d+)', function($category, $id){
    echo 'The category is: '.$category.' and the ID is: '. $id;
});

// Use the custom pattern defined above for routing;
Routex::get('/endpoint/:hostname', function($hostname){
    echo $hostname;
});

// Use predefined patterns using placeholders;
Routex::get('/endpoint/:number/another', function($number){
    echo $number;
});

// Start routing, this will return false if no route is matched and if there is no error() defined. Returns null if route is found and Closure is executed;
Routex::run();
```


### Placeholders
Routex currently support the following placeholders, these are built in patterns and then converted to regex.

- :domain = Matches any valid domain
- :number = Matches any number (digits)
- :word = Matches any word (letter, number, underscore)

### Contribute
Contributions are welcome in a form of a pull request (PR).

### License
Mozilla Public License Version 2.0


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 19-04-2022
### Fixed
- Bug that broke GET on /.

## [1.0.1] - 23-10-2021
### Fixed
- Bug that prevented routing when you use GET parameters (reported and fixed by @ayyylias).

### Changed
- License has been adjusted from Hybula Internal License to MPL-2.0.
- Minimum PHP version lowered from PHP 8.x to PHP 7.4.

### Added
- This CHANGELOG.md to track changes.
- Added support for PHP 7.4.
- Support for Composer autoloading.

## [1.0.0] - 22-09-2021
### Changed
- Routex release on GitHub.

src/Hybula/Routex/Routex.php
<?php

declare(strict_types=1);
/**
 * =========================================
 * =========================================
 * **   _  ___   _____ _   _ _      _     **
 * **  | || \ \ / / _ ) | | | |    /_\    **
 * **  | __ |\ V /| _ \ |_| | |__ / _ \   **
 * **  |_||_| |_| |___/\___/|____/_/ \_\  **
 * **                                     **
 * =========================================
 * =========================================
 *
 * Routex PHP Router (PHP)
 *
 * @package Hybula\Routex
 * @author Hybula Development Team <development@hybula.com>
 * @version 1.0.1
 * @copyright Hybula B.V.
 * @license MPL-2.0 License
 * @see https://github.com/hybula/php-routex/
 */

namespace Hybula\Routex;

class Routex
{

    private static $routes = [];
    private static $patterns = [];
    private static $error;
    private static $code = 404;
    private static $debug = false;
    private static $base = false;

    private static function pcre($pattern, $redundant = false): string
    {
        if ($redundant) {
            $pattern = ltrim($pattern, '/');
        } else {
            $pattern = trim($pattern, '/');
        }
        if (self::$base) {
            $pattern = self::$base.'/'.$pattern;
        }
        $pattern = str_replace('/', '\/', $pattern);
        foreach (self::$patterns as $placeholder => $customPattern) {
            $pattern = str_replace($placeholder, $customPattern, $pattern);
        }
        return '/^'.$pattern.'$/';
    }

    public static function base($base = false): void
    {
        self::$base = trim($base, '/');
    }

    public static function debug($debug = true): void
    {
        self::$debug = $debug;
    }

    public static function patterns(array $patterns = []): void
    {
        self::$patterns[':domain'] = '((?!\-)(?:(?:[a-zA-Z\d][a-zA-Z\d\-]{0,61})?[a-zA-Z\d]\.){1,126}(?!\d+)[a-zA-Z\d]{1,63})';
        self::$patterns[':number'] = '(\d+)';
        self::$patterns[':word'] = '(\w+)';
        self::$patterns = array_merge(self::$patterns, $patterns);
    }

    public static function get($pattern, $closure): void
    {
        self::$routes['GET'][self::pcre($pattern)] = $closure;
        self::$routes['GET'][self::pcre($pattern . '/', true)] = $closure;
    }

    public static function post($pattern, $closure): void
    {
        self::$routes['POST'][self::pcre($pattern)] = $closure;
        self::$routes['POST'][self::pcre($pattern . '/', true)] = $closure;
    }

    public static function patch($pattern, $closure): void
    {
        self::$routes['PATCH'][self::pcre($pattern)] = $closure;
        self::$routes['PATCH'][self::pcre($pattern . '/', true)] = $closure;
    }

    public static function put($pattern, $closure): void
    {
        self::$routes['PUT'][self::pcre($pattern)] = $closure;
        self::$routes['PUT'][self::pcre($pattern . '/', true)] = $closure;
    }

    public static function delete($pattern, $closure): void
    {
        self::$routes['DELETE'][self::pcre($pattern)] = $closure;
        self::$routes['DELETE'][self::pcre($pattern . '/', true)] = $closure;
    }

    public static function error($code, $closure): void
    {
        self::$code = $code;
        self::$error = $closure;
    }

    public static function load($directory): void
    {
        foreach (glob(rtrim($directory, '/') . '/*.php') as $route) {
            include $route;
        }
    }

    public static function run(): ?bool
    {
        $requestUri = ltrim(rawurldecode($_SERVER['REQUEST_URI']), '/');
        $requestUri = explode('?', $requestUri)[0];
        foreach (self::$routes as $method => $data) {
            if ($method == $_SERVER['REQUEST_METHOD']) {
                foreach ($data as $pattern => $closure) {
                    if (preg_match($pattern, $requestUri, $params)) {
                        if (self::$debug) {
                            echo 'Success ' . $requestUri . ' >>> ' . $pattern . PHP_EOL;
                        }
                        array_shift($params);
                        return call_user_func_array($closure, array_values($params));
                    }
                    if (self::$debug) {
                        echo 'Failed ' . $requestUri . ' >>> ' . $pattern . PHP_EOL;
                    }
                }
            }
        }

        http_response_code(self::$code);
        if (isset(self::$error)) {
            return call_user_func(self::$error);
        }
        return false;
    }

}
